engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/ModuleScope.java

```diff
@@ -460,4 +261,265 @@ Type getType() {
+      var allTypeMethods = methods.get(tpeKey);
+      if (allTypeMethods != null) {
+        allTypeMethods.forEach((name, fun) -> scope.registerMethod(tpeKey, name, fun));
+      }
+    }
+
+    /**
+     * Adds a dependency for this module.
+     *
+     * @param scope the scope of the newly added dependency
+     */
+    public void addImport(ImportExportScope scope) {
+      assert moduleScope == null;
+      imports.add(scope);
+    }
+
+    /**
+     * Adds an information about the module exporting another module.
+     *
+     * @param scope the exported scope
+     */
+    public void addExport(ImportExportScope scope) {
+      assert moduleScope == null;
+      exports.add(scope);
+    }
+
+    public Module getModule() {
+      return module;
+    }
+
+    /**
+     * Create a new ModuleScope.Builder which inherits from `this` `module` and `types` that need to
+     * survive the compilation.
+     *
+     * @return new ModuleScope.Builder
+     */
+    public Builder newBuilderInheritingTypes() {
+      return new Builder(this.module, new LinkedHashMap<>(this.types));
+    }
+
+    /**
+     * Materializes the builder and ensures that no further modifications to ModuleScope are
+     * possible. Action is idempotent.
+     *
+     * @return an immutable ModuleScope
+     */
+    public ModuleScope build() {
+      if (moduleScope == null) {
+        moduleScope =
+            new ModuleScope(
+                module,
+                associatedType,
+                Collections.unmodifiableMap(polyglotSymbols),
+                Collections.unmodifiableMap(types),
+                Collections.unmodifiableMap(methods),
+                Collections.unmodifiableMap(conversions),
+                Collections.unmodifiableSet(imports),
+                Collections.unmodifiableSet(exports));
+      }
+      return moduleScope;
+    }
+
+    public static ModuleScope.Builder fromCompilerModuleScopeBuilder(
+        CompilerContext.ModuleScopeBuilder scopeBuilder) {
+      return ((TruffleCompilerModuleScopeBuilder) scopeBuilder).unsafeScopeBuilder();
+    }
+
+    /**
+     * Return a view on `this` as a ModuleScope, rather than its builder.
+     *
+     * @return ModuleScope, if the builder has already been `built`, a proxy instance with the
+     *     currently registered entities
+     */
+    public ModuleScope asModuleScope() {
+      if (moduleScope != null) return moduleScope;
+      else {
+        if (proxy == null) {
+          proxy =
+              new ModuleScope(
+                  module,
+                  associatedType,
+                  Collections.unmodifiableMap(polyglotSymbols),
+                  Collections.unmodifiableMap(types),
+                  Collections.unmodifiableMap(methods),
+                  Collections.unmodifiableMap(conversions),
+                  Collections.unmodifiableSet(imports),
+                  Collections.unmodifiableSet(exports));
+        }
+        return proxy;
```

Comment:
From the method description, I thought that `asModuleScope` purpose is to make the builder re-usable by returning a _view_ on its current state.
But in fact, it is not a _view_ of the current builder's state, but a _snapshot_ taken at some point because the `proxy` is stored within the builder. I.e.
```java
var view1 = builder.asModuleScope();
builder.registerType(t);
var view2 = builder.asModuleScope();
view1 == view2; // true, but I'd expect it to be false
```

This behavior can be unexpected and lead to incorrect usage of the method.

Reply:
That is true. I was mostly concerned about having too many of such short-lived _views_. We typically get a module or an associated type and throw away the scope. But maybe it was the case of _premature_ optimization.

Reply:
> Done

Where?